opencensus-shim/src/main/java/io/opentelemetry/opencensusshim/OpenTelemetrySpanBuilderImpl.java

```diff
@@ -0,0 +1,267 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+// Includes work from:
+/*
+ * Copyright 2018, OpenCensus Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.opentelemetry.opencensusshim;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import io.opencensus.common.Clock;
+import io.opencensus.implcore.internal.TimestampConverter;
+import io.opencensus.implcore.trace.RecordEventsSpanImpl;
+import io.opencensus.implcore.trace.internal.RandomHandler;
+import io.opencensus.trace.Link;
+import io.opencensus.trace.Sampler;
+import io.opencensus.trace.Span;
+import io.opencensus.trace.Span.Kind;
+import io.opencensus.trace.SpanBuilder;
+import io.opencensus.trace.SpanContext;
+import io.opencensus.trace.SpanId;
+import io.opencensus.trace.TraceId;
+import io.opencensus.trace.TraceOptions;
+import io.opencensus.trace.Tracestate;
+import io.opencensus.trace.config.TraceConfig;
+import io.opencensus.trace.config.TraceParams;
+import java.util.Collections;
+import java.util.List;
+import java.util.Random;
+import javax.annotation.Nullable;
+
+public class OpenTelemetrySpanBuilderImpl extends SpanBuilder {
```

Comment:
Maybe I miss something, but I would see this simpler than this:

public class OpenTelemetrySpanBuilderImpl extends SpanBuilder {
  private final otel.Builder builder;

  ... // All methods redirect to the otel builder
}

public class OpenCensusSpanWrapper extends oc.Span {
  private final otel.ReadableSpan;
  private final oc.SpanContext;  // probably worth constructing this once not when getContext is called.

  ... // All methods redirect to the 
}


Can you tell me if this is possible?

Reply:
Context Interaction:
* a wrapped.Span is added to the OC context, we add the "wrapped span" to the Otel Context
* a blank oc Span is added to the OC context, we add a "blank span" to the Otel Context
* a otel.Span is added to the Otel Context, we create a wrapper and add it to OC Context.

Reply:
@bogdandrutu yep I've updated everything now, could you please take another look?

Reply:
I am not sure this is what I meant. I was trying to suggest to use composition when implementing the OC required types and delegate everything to otel.

Reply:
I'm quite confused. I'm delegating everything to oTel after creating the oTel span here: https://github.com/open-telemetry/opentelemetry-java/pull/2060/files#diff-4b913d64466c48a3f3149a7e7c9345886e02931d69396c8223fc855818f3e261R166-R167

Could you explain what you were expecting a bit more?

Reply:
I've moved some of the logic from `startSpan()` to the methods overriding OC methods, such as `addParentLinks()` and `setSpanKind()` to set Otel fields directly. The rest depend on the sampler, which does all of https://github.com/open-telemetry/opentelemetry-java/pull/2060/files#diff-4b913d64466c48a3f3149a7e7c9345886e02931d69396c8223fc855818f3e261R109-R146 to determine the kind of span that should be created. Since the sampler takes in OC parameters (parent, hasRemoteParent, traceId, spanId, name, parentLinks), I don't think I can map and delegate this nor its parameters to Otel, especially since the sampler could be implemented by the user.

Reply:
@bogdandrutu If I'm not mistaken, we can't eagerly delegate to OTel because it doesn't have `setRecordEvents` or `setSampler`. Having `setSampler` on OTel's `SpanBuilder` seems useful to me but guess it's a spec change. `setRecordEvents` we shouldn't need since pretty sure that can just be implemented as `setSampler(ALWAYS_OFF)`. If we could do it, then all the logic in this class goes away I think (would need `mapSampler` to shim the sampler)

Reply:
I see, maybe we can stick with what we have here for the moment, and look into ways to improve this.